### ISSUE TITLE
Issues with gcc/g++ detection and use on OSX

### ISSUE THREAD

**Issue by Colengms** [microsoft/vscode-cmake-tools#1653]
When I install gcc-10 on OSX (using `brew install gcc`), and `Scan for Kits`, CMake Tools will populate the `cmake-tools-kits.json` with 2 separate instances for gcc.   The one titled "GCC 10.2.0", which seems valid.  Another titled "GCC for x86_64-apple-darwin20-gcc-10.2.0" contains a "C" compiler but no CPP compiler.

```
{
    "name": "GCC for x86_64-apple-darwin20 10.2.0",
    "compilers": {
        "C": "/usr/local/bin/x86_64-apple-darwin20-gcc-10.2.0".
    }
}
```

There is a `g++` symlink present in the same folder, titled, `x86_64-apple-darwin20-g++10`, which reports a 10.2.0 version, so appears to be the matching g++.  Oddly, these appear to be separate distinct binaries (and not symlinks to the same binary used by the "GCC 10.2.0" configuration).  If it's desirable for both kits to be detected, it should probably pick up the corresponding g++ path for this one, otherwise the kit does not work.  With both C and C++ files, I see CMake Tools falls back to Apple Clang.

Also, when configuring CMake Tools with the valid "GCC 10.2.0" kit, I receive a browse config containing a `compilerPath` of `/usr/bin/clang++`.  Shouldn't CMake Tools be using the compiler specified in the selected kit for the `compilerPath` of the browse config?

**Comment by andreeis** [comment]
I see the code in CMake Tools is written on the assumption that there would be a g++ with the exact same suffix/prefix as its gcc counterpart. In this case, unless there was a typo when you wrote the paths, for some reason the g++ one doesn't have the -10.2.0 ending. The extension replaces gcc with g++ and searches for the file. Of course, we can improve on this to accept more varied situations of tools lying around.

I am not able to reproduce yet your second bug, but I also didn't create the exact situation you mention. I switch between 2 valid kits and verify under the debugger the compiler path. But I did notice that for it to change, I have to do a reload. So one bug I can look at is why it doesn't get updated after the kit is changed. In your case, do you see a correct compiler path after you reload after you change back to the valid kit?

**Comment by Colengms** [comment]
>unless there was a typo when you wrote the paths

I did not specify any paths.  I just ran "brew install gcc".  That is what it created by default.  There is also a binary for gcc without the "10.2.0" present in the filename, in the same folder, which was not detected.

**Comment by Colengms** [comment]
Note that the compiler path I was referring to in the 'second bug' was from the 'browse config', not a configuration for a specific file.  I can repro evidence of this by:

 * enabling `"C_Cpp.loggingLevel": "Debug"`
 * closing all source files
 * freshly launching VS Code
 * running a C/C++ Extension command such as Log Diagnostics to make sure it's activated
 * clearing out all log output from cpptools
 * Selecting the "GCC 10.2.0" kit
 * Running `CMake: Delete Cache and Reconfigure`.

The output of the C/C++ Extension should now refer only to the browse config, and no file configs.  Although we don't log the browse config object currently, I am seeing that the occurance of `cpptools/didChangeCustomBrowseConfiguration` is immediately following by an attempt to query "/usr/bin/clang++" for compiler defaults.

I locally added logging of the browse config, and I can see that /usr/bin/clang++ was reported in this case, and not the compiler from the selected kit.

**Comment by andreeis** [comment]
Thanks for the extra repro information. We will address both (I may split this in two later). Still one question, after you reproduce the symptom, reload the project in VSCode. Is everything configured properly now or not? I want to confirm if reloading the project can be an acceptable workaround.

**Comment by Colengms** [comment]
@andreeis No, if I follow the steps in my previous comment, reload VS Code again (with no source files open), and run `CMake: Configure`, I see the same behavior (wrong compiler used in browse config).

**Comment by stertingen** [comment]
This seems to affect also Linux/WSL with custom toolchains.

I have a custom toolchain installed at `/opt/gcc-linaro-7.2.1-2017.11-x86_64_arm-linux-gnueabihf`. The kit scan only detects `/opt/gcc-linaro-7.2.1-2017.11-x86_64_arm-linux-gnueabihf/bin/arm-linux-gnueabihf-gcc-7.2.1` for the C compiler, but not the C++ compiler, which does not have the version suffix.

**Comment by bobbrow** [comment]
The fix for this issue is available in 1.10 "pre-release".  You can help us validate it by:
1. Opening the extensions panel in VS Code
2. Selecting the CMake Tools extension
3. Clicking the "Switch to Pre-release Version" button

![image](https://user-images.githubusercontent.com/12818240/156218538-7bce4a1f-4fc9-43cc-afb0-7f423aa5980e.png)
